go-controller/pkg/ovn/master.go

```diff
@@ -842,31 +848,122 @@ func (oc *Controller) addNodeAnnotations(node *kapi.Node, hostSubnets []*net.IPN
+	// or that it needs to allocate new subnet because it is a new node
+	// or has been converted from single-stack to dual-stack
+	klog.Infof("Expected %d subnets on node %s, found %d: %v",
+		expectedHostSubnets, node.Name, currentHostSubnets, hostSubnets)
+	// release unexpected subnets
+	// filter in place slice
+	// https://github.com/golang/go/wiki/SliceTricks#filter-in-place
+	foundIPv4 := false
+	foundIPv6 := false
+	n := 0
+	for _, subnet := range hostSubnets {
+		// if the subnet is not going to be reused release it
+		if config.IPv4Mode && utilnet.IsIPv4CIDR(subnet) && !foundIPv4 {
+			klog.V(5).Infof("Valid IPv4 allocated subnet %v on node %s", subnet, node.Name)
+			hostSubnets[n] = subnet
+			n++
+			foundIPv4 = true
+			continue
+		}
+		if config.IPv6Mode && utilnet.IsIPv6CIDR(subnet) && !foundIPv6 {
+			klog.V(5).Infof("Valid IPv6 allocated subnet %v on node %s", subnet, node.Name)
+			hostSubnets[n] = subnet
+			n++
+			foundIPv6 = true
+			continue
+		}
+		// this subnet is no longer needed
+		klog.V(5).Infof("Releasing subnet %v on node %s", subnet, node.Name)
+		_ = oc.masterSubnetAllocator.ReleaseNetwork(subnet)
+	}
+	// recreate hostSubnets with the valid subnets
+	hostSubnets = hostSubnets[:n]
+	// allocate new subnets if needed
+	if config.IPv4Mode && !foundIPv4 {
+		allocatedHostSubnet, err := oc.masterSubnetAllocator.AllocateIPv4Network()
+		if err != nil {
+			return nil, nil, fmt.Errorf("error allocating network for node %s: %v", node.Name, err)
+		}
+		// the allocator returns nil if it can't provide a subnet
+		// we should filter them out or they will be appended to the slice
+		if allocatedHostSubnet != nil {
+			klog.V(5).Infof("Allocating subnet %v on node %s", allocatedHostSubnet, node.Name)
+			allocatedSubnets = append(allocatedSubnets, allocatedHostSubnet)
+			// Release the allocation on error
+			defer func() {
+				if err != nil {
+					klog.Warningf("Releasing subnet %v on node %s: %v", allocatedHostSubnet, node.Name, err)
+					_ = oc.masterSubnetAllocator.ReleaseNetwork(allocatedHostSubnet)
+				}
+			}()
+		}
+	}
+	if config.IPv6Mode && !foundIPv6 {
+		allocatedHostSubnet, err := oc.masterSubnetAllocator.AllocateIPv6Network()
+		if err != nil {
+			return nil, nil, fmt.Errorf("error allocating network for node %s: %v", node.Name, err)
+		}
+		// the allocator returns nil if it can't provide a subnet
+		// we should filter them out or they will be appended to the slice
+		if allocatedHostSubnet != nil {
+			klog.V(5).Infof("Allocating subnet %v on node %s", allocatedHostSubnet, node.Name)
+			allocatedSubnets = append(allocatedSubnets, allocatedHostSubnet)
+		}
+	}
+	// check if we were able to allocate the new subnets required
+	wantedSubnets := expectedHostSubnets - currentHostSubnets
+	if wantedSubnets > 0 && len(allocatedSubnets) != wantedSubnets {
+		return nil, nil, fmt.Errorf("error allocating networks for node %s: %d subnets expected only new %d subnets allocated",
```

Comment:
I think that if we end up here: we won't release the allocated subnet in the assigned `defer` function. You're missing setting the variable `err` here. I think that what you need is:
```
err := fmt.Errorf("error allocating networks for node %s: %d subnets expected only new %d subnets allocated", node.Name, expectedHostSubnets, len(allocatedSubnets))
return nil, nil, err
```

Reply:
Or maybe you don't want to release the allocated subnet in this case? But in that case we won't proceed with any other setup in `addNode` either.

Reply:
this is something I had implement because of the allocators behaviour, they don't fail if they can´t allocate a network, they just return nil.
So, the only way I see we fail here is that the allocators didn't allocated the subnets, maybe we should be pedantic ...

Reply:
I remember now, this case is only possible if the user has set a wrong configuration, see the test case
https://github.com/ovn-org/ovn-kubernetes/pull/2013/files#diff-da0dab63258c018f29107af5cc19fdfce50a8b86d24d9a2e285fff4334187845R1419-R1433
In that case we don't have to release anything because the user has to reconfigure the cluster